bug: Unable to convert YAML program to any other language after version 3.87.0

### What happened?

Trying to convert a YAML program to Typescript:

```bash
pulumi convert --generate-only --language typescript  --out ../typescript/
```

`pulumi` crashes with the following panic:

```
Converting from yaml...
warning: using pulumi-resource-mssql from $PATH at /home/vagrant/work/pulumi-mssql/bin/pulumi-resource-mssql
warning: using pulumi-resource-mssql from $PATH at /home/vagrant/work/pulumi-mssql/bin/pulumi-resource-mssql
Converting to nodejs...
panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0xe814b7]

goroutine 33 [running]:

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.(*binder).outputVersionSignature(0xdf39e5?, 0xc000133f80)

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/invoke.go:259 +0x57

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.(*binder).signatureForArgs(0xc0004cc5a0?, 0x1a?, {0x24a2560?, 0xc000957d40?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/invoke.go:198 +0x10c

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.(*binder).bindInvokeSignature(0xc000133700, {0xc000a68040, 0x2, 0xc000957d40?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/invoke.go:155 +0x1553

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.GenericFunctionSignature.GetSignature(0xc0006f1d88?, {0xc000a68040?, 0xcbf725?, 0x2487870?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/functions.go:59 +0x4b

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.(*Function).GetSignature(...)

        /home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/functions.go:84

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.(*expressionBinder).bindFunctionCallExpression(0xc000ca2ce8, 0xc000029c20)

        /home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/binder_expression.go:438 +0x2eb

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.(*expressionBinder).bindExpression(0xc000ca2d28?, {0x2487730?, 0xc000029c20?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/binder_expression.go:91 +0xc5

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.BindExpression({0x2487730, 0xc000029c20}, 0xc00029ac30, {0x24875f0, 0xc00029ac00}, {0xc000a624c0, 0x2, 0xc000ca2e00?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/binder_expression.go:65 +0xdf

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.BindAttribute(0xc000170380, 0x0?, {0x24875f0, 0xc00029ac00}, {0xc000a624c0, 0x2, 0x2})

        /home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/attribute.go:82 +0x9c

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.(*binder).bindLocalVariable(0xc000133700, 0xc000770900)

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/binder_nodes.go:164 +0x132

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.(*binder).bindNode(0xc000133700?, {0x24a1708, 0xc000770900})

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/binder_nodes.go:59 +0x485

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.(*binder).bindNode(0x1ab0b40?, {0x24a1618, 0xc0007560b0})

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/binder_nodes.go:50 +0x2ff

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.BindProgram({0xc0008f0188?, 0x1, 0x1}, {0xc00045b810, 0x9, 0x0?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/binder.go:219 +0xd05

github.com/pulumi/pulumi/pkg/v3/codegen/pcl.BindDirectory({0xc0008c2240, 0x1d}, {0x2487d48?, 0xc000ae4990}, {0xc00029a6f0, 0x6, 0xc0008fc840?})

        /home/runner/work/pulumi/pulumi/pkg/codegen/pcl/binder.go:280 +0x5b6

main.(*nodeLanguageHost).GenerateProject(0x0?, {0xc0005417d8?, 0x4105c5?}, 0xc000132b00)

        /home/runner/work/pulumi/pulumi/sdk/nodejs/cmd/pulumi-language-nodejs/main.go:1033 +0x1ca

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler.func1({0x2497550, 0xc00029a3f0}, {0x1e04100?, 0xc000132b00})

        /home/runner/work/pulumi/pulumi/sdk/proto/go/language_grpc.pb.go:445 +0x72

github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x2497550, 0xc0008e5ad0}, {0x1e04100, 0xc000132b00}, 0xc0009b05e0, 0xc000ae4948)

        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x3d0

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler({0x1e56fa0?, 0xc000836140}, {0x2497550, 0xc0008e5ad0}, 0xc000458620, 0xc0009b0080)

        /home/runner/work/pulumi/pulumi/sdk/proto/go/language_grpc.pb.go:447 +0x135

google.golang.org/grpc.(*Server).processUnaryRPC(0xc0008f2000, {0x249f7b0, 0xc000708ea0}, 0xc0000fcfc0, 0xc0008e43c0, 0x33af870, 0x0)

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:1360 +0xe15

google.golang.org/grpc.(*Server).handleStream(0xc0008f2000, {0x249f7b0, 0xc000708ea0}, 0xc0000fcfc0, 0x0)

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:1737 +0x9e7

google.golang.org/grpc.(*Server).serveStreams.func1.1()

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:982 +0x8d

created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 30

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:980 +0x165

error: could not generate output program: error reading from server: EOF
```

### Example

```yaml
name: mssql-test
runtime: yaml
description: A Pulumi YAML project to test the new MSSQL provider

variables:
  location: "eastus2"
  current:
    fn::invoke:
      function: azure:core:getClientConfig
      options:
        provider: ${provider}

  azuread-group-database-administrators:
    fn::invoke:
      function: azuread:getGroup
      arguments:
        displayName: "Database Administrators"
        securityEnabled: true
      options:
        provider: ${provider-azuread}

  mssql-database:
    fn::invoke:
      function: mssql:getDatabase
      arguments:
        name: ${database.name}
      options:
        provider: ${provider-mssql}

  mssql-database-role-owner:
    fn::invoke:
      function: mssql:getDatabaseRole
      arguments:
        databaseId: ${mssql-database.id}
        name: "db_owner"
      options:
        provider: ${provider-mssql}


outputs: {}

resources:
  provider:
    type: pulumi:providers:azure
    properties:
      skipProviderRegistration: true
    options:
      version: 5.52.0

  provider-azuread:
    type: pulumi:providers:azuread
    options:
      version: 5.0.0

  random-id:
    type: random:RandomId
    properties:
      keepers:
        keep: ${location}
      byteLength: 8

  resource-group:
    type: azure:core:ResourceGroup
    properties:
      name: "pulumi-mssql-test-${random-id.hex}"
      location: ${location}

  server:
    type: azure:mssql:Server
    properties:
      name: "pulumi-mssql-test-${random-id.hex}"
      resourceGroupName: ${resource-group.name}
      location: ${resource-group.location}
      version: '12.0'
      minimumTlsVersion: '1.2'
      azureadAdministrator:
        loginUsername: "pulumi-app"
        objectId: ${current.objectId}
        tenantId: ${current.tenantId}
        azureadAuthenticationOnly: true
    options:
      provider: ${provider}

  database-firewall-rule:
    type: azure:mssql:FirewallRule
    properties:
      name: azure
      serverId: ${server.id}
      startIpAddress: "0.0.0.0"
      endIpAddress: "0.0.0.0"
    options:
      provider: ${provider}

  database:
    type: azure:mssql:Database
    properties:
      name: "db"
      serverId: ${server.id}
      licenseType: LicenseIncluded
      maxSizeGb: 2
      skuName: Basic
    options:
      provider: ${provider}

  provider-mssql:
    type: pulumi:providers:mssql
    properties:
      hostname: ${server.fullyQualifiedDomainName}
      azureAuth: {}
    options:
      dependsOn:
        - ${database}
        - ${database-firewall-rule}

  mssql-database-admin:
    type: mssql:AzureadUser
    properties:
      name: "Database Administrators"
      databaseId: ${mssql-database.id}
      userObjectId: ${azuread-group-database-administrators.objectId}
    options:
      provider: ${provider-mssql}

  mssql-database-administrators-member:
    type: mssql:DatabaseRoleMember
    properties:
      roleId: ${mssql-database-role-owner.id}
      memberId: ${mssql-database-admin.id}
    options:
      provider: ${provider-mssql}

```

### Output of `pulumi about`

CLI          
Version      3.87.0
Go Version   go1.21.1
Go Compiler  gc

Plugins
NAME     VERSION
azure    5.52.0
azuread  5.0.0
mssql    unknown
random   unknown
yaml     unknown

Host     
OS       ubuntu
Version  22.04
Arch     x86_64

This project is written in yaml

Current Stack: organization/mssql-test/default

Found no resources associated with default

Found no pending operations associated with default

Backend        
Name           pulumi
URL            file://.stacks
User           vagrant
Organizations  
Token type     personal

No dependencies found

### Additional context

Checked that the latest version of pulumi that converts the YAML program successfully is: 3.86.0

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20231012--programgen--fixes-panic-when-binding-the-signature-of-output-versioned-invokes-without-input-arguments.yaml
+++ b/changelog/pending/20231012--programgen--fixes-panic-when-binding-the-signature-of-output-versioned-invokes-without-input-arguments.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen
+  description: Fixes panic when binding the signature of output-versioned invokes without input arguments 

--- a/pkg/codegen/pcl/invoke.go
+++ b/pkg/codegen/pcl/invoke.go
@@ -255,8 +255,13 @@ func (b *binder) outputVersionSignature(fn *schema.Function) (model.StaticFuncti
 		return model.StaticFunctionSignature{}, fmt.Errorf("Function %s does not have an Output version", fn.Token)
 	}
 
-	// Given `fn.NeedsOutputVersion()==true`, can assume `fn.Inputs != nil`, `fn.ReturnType != nil`.
-	argsType := b.schemaTypeToType(fn.Inputs.InputShape)
+	// Given `fn.NeedsOutputVersion()==true` `fn.ReturnType != nil`.
+	var argsType model.Type
+	if fn.Inputs != nil {
+		argsType = b.schemaTypeToType(fn.Inputs.InputShape)
+	} else {
+		argsType = model.NewObjectType(map[string]model.Type{})
+	}
 	returnType := b.schemaTypeToType(fn.ReturnType)
 	return b.makeSignature(argsType, model.NewOutputType(returnType)), nil
 }

